### DISCUSSION TITLE
fixEff in RRBLUP() function

### DISCUSSION THREAD

**Discussion by marcopxt** [gaynorr/AlphaSimR#115] [Q&A]
Hey, all!

So, when I was looking at how to use fixed effects into the RRBLUP() function, the manual says that the fixEff from a population can be stored and used latter by the genomic selection models:
'fixEff: a fixed effect relating to the phenotype. Used by genomic selection models but otherwise ignored.'

However, there is no argument into RRBLUP() or its variations that sets the fixEff from the populations into the model. That being said, it always accounts for the fixEff from the population or am I missing something?

Thank you,

**Comment by gaynorr** [comment]
The fixEff slot gets set by the `setPheno` function. The main use for it is to handle genomic selection in simulations that are modeling GxE. The idea is to have a different level for the fixed effect for each environment. 

You can also manually change the values in the fixEff slot. The values are integers that get treated as levels of a factor. These values get used within the C++ side of the RRBLUP function to create a sum-to-zero design matrix for the fixed effects.

I've typically use the GxE functionality to model genotype-by-year interactions in the simulations I'm running. When this the case, I usually pass the current year as the value of the fixed effect. I'll post a link to an example script in the next day or two. Feel free to ping me if I forget.

**Reply by marcopxt** [reply]
Thank you for your reply!

If the populations used for the genomic model have the fixEff set to different values it will be automatically read and used by the RRBLUP() function?

I am saying this once there is no argument that refers to fixEff into RRBLUP function.

Thank you

**Reply by gaynorr** [reply]
Yes. I often build training populations by concatenating multiple populations together. In this way I can combine yield trials from different years into one training population to better reflect how GS is used in a plant breeding program.

**Comment by gaynorr** [comment]
Here's an old example I had on my computer showing how to use the fixed effects in AlphaSimR. There's not much documentation, so please ask questions if anything isn't clear. I've been meaning to write some of this up in a vignette. Maybe one day I'll finally make it happen.

```
library(AlphaSimR)

founderPop = quickHaplo(100, 10, 100)

SP = SimParam$
  new(founderPop)$
  restrSegSites(overlap=TRUE)$ # Allowing QTL and SNP to overlap
  addTraitA(100)$
  setVarE(h2=0.3)$
  addSnpChip(100)

pop = newPop(founderPop)

# Generate populations for 4 locations
loc1 = setPheno(pop, fixEff=1)
loc2 = setPheno(pop, fixEff=2)
loc3 = setPheno(pop, fixEff=3)
loc4 = setPheno(pop, fixEff=4)

# # Manually adding fixed effects. This is needed if
# # fixed effects aren't already added using the setPheno 
# # function as done above. This is need if phenotypes 
# # are generated in some other way.
# loc1@fixEff = rep(1L, nInd(loc1))
# loc2@fixEff = rep(2L, nInd(loc2))
# loc3@fixEff = rep(3L, nInd(loc3))
# loc4@fixEff = rep(4L, nInd(loc4))

# Combine locations into a training population
trainPop = c(loc1, loc2, loc3, loc4)

# Run GS model
model = RRBLUP(trainPop)

# Set EBV
pop = setEBV(pop, model)

# Measure accuracy
cor(gv(pop), ebv(pop))

### Regarding the fixed effects and GS model
# The fixed effects get treated as a categorical
# variable and are used to construct a sum-to-zero
# design matrix. AlphaSimR actually does this in the 
# C++ code, but here is an equivalent R code for 
# demonstration. When you have years and locations, 
# you want a unique fixed effect for each year by location 
# combination. 
fixedEff = as.factor(trainPop@fixEff)
X = model.matrix(~fixedEff, 
                 contrasts=list(fixedEff="contr.sum"))
```

**Reply by marcopxt** [reply]
Awesome! That answer all my doubts!

Thank you so much for your time and for the code :)
